text/0300-programmatic-toolkit.md

```diff
@@ -0,0 +1,745 @@
+Providing a function allows your implementation to integrate with the [Programmatic Toolkit]'s multi-pass requirement.
+The function will be called one or more times with the known context values of the current iteration.
+It must then use these values to synthesize and return a `CloudAssembly`.
+For all features (e.g. lookups) to work correctly, the `App()` must be instantiated with the received context values.
+Since it is not possible to update the context of an app, it must be created inside the function scope.
+
+A basic implementation would look like this:
+
+```ts
+const cx: ICloudExecutable = CloudExecutable.fromCloudAssemblyProducer(
+  async (context: Record<string, any>): cxapi.CloudAssembly => {
+    const app = new cdk.App({ context });
+    const stack = new cdk.Stack(app);
+    return app.synth();
+  }
+);
+```
+
+Existing applications, can also be used by simply passing an AWS CDK app directory (containing a `cdk.json` file):
+
+```ts
+const cx: ICloudExecutable = CloudExecutable.fromCdkApp("/path/to/cdk/app");
+```
+
+Alternatively the `cdk.out` directory from a previous synth can be used as well:
+
+```ts
+const cx: ICloudExecutable = CloudExecutable.fromCdkOut("/path/to/cdk.out");
+```
+
+#### Actions - synth
+
+The `synth` action takes a Cloud Executable and synthesizes it in to a `CloudAssembly`.
+
+```ts
+const cx: ICloudExecutable =  await cdk.synth(cx, { stacks: ['MyTestStack'] });
+```
+
+It returns a Cloud Executable that is cached on disk and can be used as input for other actions.
+
+#### Actions - bootstrap
+
+The `bootstrap` action takes in a list of environments and bootstraps each environment with the provided configuration:
+
+```ts
+const envs: string[] = cx.environments();
+await cdk.bootstrap(envs, { trust: ['000000000000', '111111111111'] });
+```
+
+It returns a list of stack deployments.
+
+#### Errors
+
+All errors are thrown as exceptions and the integrator must be prepared to handle them.
+Expected errors are structured data and extend the `IoMessage` interface.
+
+```ts
+interface ToolkitError extends IoMessage {
+  source: 'toolkit' | 'user';
+  type: 'authentication' | 'cloudformation' | 'validation' ...;
+  fqn?: string;
+}
+```
+
+The toolkit might throw other exceptions.
+These are bugs and you should report them by [raising an issue](https://github.com/aws/aws-cdk/issues/new?assignees=&labels=bug%2Cneeds-triage&projects=&template=bug-report.yml&title=%28toolkit%29%3A+Untyped+Exception+%28short+description%29).
+To assist integrators with detecting the type of error, the following helper methods are available.
+Even though errors are typed, you should not rely on `instanceof` checks because it can behave unexpectedly when working with multiple copies of the same package. [TODO: insert reference link]
+
+```ts
+try {
+  await cdk.deploy(...);
+} catch (e: unknown) {
+  if (ToolkitError.isToolkitError(error)) {
+    // handle toolkit errors
+  } else if (ToolkitError.isUserError(e)) {
```

Comment:
What does this mean? You just said if we throw anything other than a `ToolkitError` it's a bug.